macOS/unwindmacos.cpp

```diff
@@ -0,0 +1,210 @@
+    SecondLevelLookupCompressed(module, original_address, translated_address, first_level_entry, second_level_page_addr);
+  } else if (*(uint32_t*)second_level_page_addr == UNWIND_SECOND_LEVEL_REGULAR) {
+    SecondLevelLookupRegular(module, original_address, translated_address, first_level_entry, second_level_page_addr);
+  }
+}
+
+
+void UnwindGeneratorMacOS::SecondLevelLookupCompressed(ModuleInfo *module,
+                                                       size_t original_address,
+                                                       size_t translated_address,
+                                                       struct unwind_info_section_header_index_entry *first_level_entry,
+                                                       size_t second_level_page_addr) {
+  struct unwind_info_compressed_second_level_page_header *second_level_header =
+    (struct unwind_info_compressed_second_level_page_header *)second_level_page_addr;
+
+  uint32_t second_level_entry = 0;
+  size_t second_level_entries_start = second_level_page_addr + second_level_header->entryPageOffset;
+
+  uint16_t low_index = 0;
+  uint16_t high_index = second_level_header->entryCount;
+  while (low_index < high_index) {
+    uint16_t mid_index = low_index + (high_index - low_index) / 2;
+    size_t curr_second_level_entry_addr = second_level_entries_start + mid_index * sizeof(uint32_t);
+
+    uint32_t curr_entry = *(uint32_t*)curr_second_level_entry_addr;
+    uint32_t curr_entry_encoding_index = UNWIND_INFO_COMPRESSED_ENTRY_ENCODING_INDEX(curr_entry);
+    uint32_t curr_entry_func_offset = UNWIND_INFO_COMPRESSED_ENTRY_FUNC_OFFSET(curr_entry);
+
+    if ((size_t)module->module_header + first_level_entry->functionOffset + curr_entry_func_offset <= original_address) {
+      second_level_entry = curr_entry;
+      low_index = mid_index + 1;
+    } else {
+      high_index = mid_index;
+    }
+  }
+
+  compact_unwind_encoding_t encoding;
+  uint32_t entry_encoding_index = UNWIND_INFO_COMPRESSED_ENTRY_ENCODING_INDEX(second_level_entry);
+  uint32_t entry_func_offset = UNWIND_INFO_COMPRESSED_ENTRY_FUNC_OFFSET(second_level_entry);
+  if (entry_encoding_index < module->unwind_data->header->commonEncodingsArrayCount) {
+    encoding = *(compact_unwind_encoding_t*)((size_t)module->unwind_data->buffer
+                                             + module->unwind_data->header->commonEncodingsArraySectionOffset
+                                             + entry_encoding_index * sizeof(compact_unwind_encoding_t));
+  } else {
+    encoding = *(compact_unwind_encoding_t*)((size_t)module->unwind_data->buffer
+                                             + second_level_header->encodingsPageOffset
+                                             + (entry_encoding_index - module->unwind_data->header->commonEncodingsArrayCount) * sizeof(compact_unwind_encoding_t));
+  }
+
+  module->unwind_data->AddEncoding(encoding, translated_address);
```

Comment:
The names of the methods such as `FirstLevelLookup`, `SecondLevelLookup` suggest to me as the reader that they are only responsible for performing look-ups without modifying state, so it's a bit counter-intuitive to see the `AddEncoding()` call here. Optionally, maybe the "lookup" methods could be made to just return the encoding, and their caller itself would be responsible for adding it to the `unwind_data` - such as `OnBasicBlockStart`, or a new helper method named e.g. `AddEncodingForAddress` or something like that.